### ISSUE TITLE
Fast Attention TF: Unit test unstable?

### ISSUE THREAD

**Issue by mschoeler** [google-research/google-research#503]
When executing `test_softmax_noncausal_attention_block_output` in `performer/fast_attention/tensorflow/fast_attention_test.py`
it produces a maximum error of 0.20 which is clearly below the max allowed error `max_error = 2.0`.
However, when changing `length` to a different value. The maximum relative error jumps up significantly.
The commit I'm running on is: 6f29c8099fbb1474769046276c8744521470cab9
I run on TF 1.15.
Here is a table of length vs error and unit test state:

length | relative error | unit test status
------|-----|------
2 | 0.20 | Pass
4 | 1.24 | Pass
8 | 106 | FAIL
10 | 9.7 | FAIL

Ping @xingyousong

**Comment by mschoeler** [comment]
Also confirmed with TF version 2.4.0

**Comment by kchoro** [comment]
How many random features do you use ? Also, we recently updated our TF implementation (see: https://github.com/google-research/google-research/commit/b09ac837cd5720bc60f1c16b472a7ab462b0ddb8) (previously there was a bug in the code).  Did you run the most updated version ?

**Comment by kchoro** [comment]
Also, I think what you observe (for some values of L) is simply much larger *relative* error. The way we compute it in unit tests is: |our_value - exact_value|/exact_value per entry of a matrix (abd then take max over all entries). So if exact value is very small, even with very good estimator (with variance going to 0 as approximated values go to 0) you might have large relative error and still very small absolute error which is really what matters at the end. We will update the unit tests to measure absolute error since it is more meaningful and and informative than the relative error.

**Comment by mschoeler** [comment]
Hi @kchoro,
the commit b09ac83 is a predecessor of 6f29c80, so I should have those changes included.
Here is the configuration, which is the same as the standard, except varying in `length`
```
batch_size = 1
length = VAR
num_heads = 1
dim = 8
num_random_features = 30000
```

Your argument makes definetely sense. I updated the table with the max absolute and mean absolute errors.
While the mean error seems to 


I used the following code to generate the errors (mostly the unit test code):
```python
def test_softmax_noncausal_attention_block_output(self):
    batch_size = 1
    length = 50 # <-- this varibale is changed for the table
    num_heads = 1
    dim = 8
    num_random_features = 30000
    query = tf.random.normal([batch_size, length, num_heads, dim])
    key = tf.random.normal([batch_size, length, num_heads, dim])
    value = tf.random.normal([batch_size, length, num_heads, dim])
    kernel_transformation = fast_attention.softmax_kernel_transformation
    projection_matrix = fast_attention.create_projection_matrix(
        num_random_features, dim)
    attention_block_output = fast_attention.favor_attention(
        query, key, value, kernel_transformation, False, projection_matrix)
  
    query = tf.multiply(query, 1.0 / math.sqrt(float(dim)))
    attention_scores = tf.einsum("BXHD,BYHD->BXYH", query, key)
    attention_scores = tf.nn.softmax(attention_scores, axis=2)
    exact_attention_block_output = tf.einsum("BXYH,BYHD->BXHD",
                                             attention_scores, value)
    error = tf.math.abs(exact_attention_block_output - attention_block_output)
    print("Max: {}".format(tf.math.reduce_max(error)))
    print("Mean: {}".format(tf.math.reduce_mean(error)))
```

length | max abs_error | mean abs_error
---- | ----- | ----
2 | 0.081 | 0.023
4 | 0.19 | 0.047
10 | 0.20 | 0.064
50 | 0.67 | 0.045
500 | 0.67 | 0.022

This is a scatter plot showing abs_error `error` over `exact_attention_block_output`. The error sometimes seems to be quite big
![Figure_1](https://user-images.githubusercontent.com/3592251/103022362-1e1d4d80-454c-11eb-811d-9e2d83cc52f3.png)

**Comment by kchoro** [comment]
Hi @mschoeler, thank you for an update. What I see is that the mean absolute error is pretty small which is great. My guess is that the small fraction of larger values of absolute error (most of the values are concentrated around 0) correspond also to larger values that are being approximated. Also take into account that the computational block includes renormalization so it is not purely an approximation of the linear transformation of the attention matrix.

**Comment by mschoeler** [comment]
Thanks for your quick reply. 
The error could also be heavily influenced by the dimensionality of the values. Since the weighted sum becomes more similar, if values are sampled from a lower dimensional space I assume.
Wouldn't it make sense to add a unit test for the kernel approximation of the softmax directly?

**Comment by kchoro** [comment]
Hi @mschoeler, very good point. Yes, we will definitely add this test to the open-sourced version. We have it in our internal codebase and as you noticed, indeed the error is even smaller if we focus only on the attention matrix. Of course the test can be implemented in particular by taking the value matrix to be an identity.

**Comment by mschoeler** [comment]
I guess to mimic this, the value matrix would need to be identity and `d_value == length`.
So basically a quadratic matrix with main diagonal set to 1

**Comment by kchoro** [comment]
Yes, exactly (see my comment above, I guess we posted our comments at the same time :))

**Comment by mschoeler** [comment]
👍

**Comment by mschoeler** [comment]
@kchoro I actually tested our proposal in the unit test.
So I created V as an identity matrix in order to be able to see the attention scores in the performer.
Here is the code:
```python
def test_softmax_noncausal_attention_block_output(self):
        batch_size = 1
        length = 32
        num_heads = 1
        dim = 32
        num_random_features = 2000
        tf.random.set_random_seed(5)
        query = tf.random.normal([batch_size, length, num_heads, dim])
        key = tf.random.normal([batch_size, length, num_heads, dim])
        value = tf.eye(dim, batch_shape=[batch_size, num_heads])
        value = tf.transpose(value, perm=[0, 2, 1, 3])
        kernel_transformation = softmax_kernel_transformation
        projection_matrix = create_projection_matrix(
            num_random_features, dim)
        attention_block_output = favor_attention(
            query, key, value, kernel_transformation, False, projection_matrix)


        ## Check that attention weights for performer sum to one
        sum_of_weights = tf.math.reduce_sum(attention_block_output,axis=-1)
        self.assertLess(tf.abs(tf.reduce_max(sum_of_weights) - 1 ), 10e-4 )

        ## Check difference to exact attention
        query = tf.multiply(query, 1.0 / math.sqrt(float(dim)))
        attention_scores = tf.einsum("BXHD,BYHD->BXYH", query, key)
        attention_scores = tf.nn.softmax(attention_scores, axis=2)
        exact_attention_block_output = tf.einsum("BXYH,BYHD->BXHD",
                                                 attention_scores, value)

        ## Visualize attention scores
        from matplotlib import pyplot as plt
        fig, (ax0, ax1, ax2) = plt.subplots(nrows=3, constrained_layout=True)
        im0 = ax0.imshow(attention_scores[0,:,:,0])
        ax0.set_title("attention_scores")

        im1 = ax1.imshow(exact_attention_block_output[0, :, 0, :])
        ax1.set_title("exact_attention_block_output")
        im2 = ax2.imshow(attention_block_output[0, :, 0, :])
        ax2.set_title("attention_block_output")
        fig.colorbar(im0, ax=ax0)
        fig.colorbar(im1, ax=ax1)
        fig.colorbar(im2, ax=ax2)
        plt.show()
```
While the fast attention scores are normalized for a certain query the visual comparisong of the attention scores between fast_attention and the exact attention show quite some differences.

![FastAttentionScoreCompare](https://user-images.githubusercontent.com/3592251/107960707-3be3bd80-6fa5-11eb-8882-7a015cd4ac0b.png)
Top: Exact attention scores; Middle: Exact attention block output; Bottom: FastAttention block output

Any idea, what could cause that discrepancy?
